src/zeroband/inference/pipeline.py

```diff
@@ -0,0 +1,221 @@
+
+
+def setup_comm(world_size: int, iroh_seed: int | None, iroh_peer_id: str | None) -> Node:
+    """
+    Setup communication via PRIME-IROH. Forms a ring topology between the model shards
+    with unidirectional communication flow.
+
+    Args:
+        world_size: The total number of model shards
+        iroh_seed: The seed for the PRIME-IROH node (optional, will lead to deterministic connection strings)
+        iroh_peer_id: The peer ID for the PRIME-IROH node (optional, will connect to a random peer if not provided)
+    """
+    assert world_size > 1, "Pipeline parallel inference requires at least 2 stages"
+
+    # Setup node (with or without seed)
+    if iroh_seed is not None:
+        # If seed is provided, create a new node with the seed
+        node = Node.with_seed(num_streams=1, seed=iroh_seed)
+    else:
+        # If no seed, create a new node
+        node = Node(num_streams=1)
+    logger.info(f"Created node (ID={node.node_id()})")
+
+    # Connect to peer
+    if iroh_peer_id is None:
+        iroh_peer_id = input("Enter Peer ID: ").strip()
+    logger.info(f"Setting up outgoing connection to {iroh_peer_id}")
+    node.connect(iroh_peer_id)
+    logger.info(f"Outgoing connection to {iroh_peer_id} successful!")
+
+    # Wait for connection to sender and receiver to be established
+    # Note: This requires the PP communication loop to be closed, e.g. for 4 stages:
+    # 0 -> 1 -> 2 -> 3 -> 0
+    logger.info("Waiting for incoming connection...")
+    while not node.is_ready():
+        time.sleep(0.1)
+    logger.info("All connections successful!")
+
+    return node
+
+
+def setup_hooks(rank: int, world_size: int, llm: LLM, node: Node) -> None:
+    """
+    Setup hooks to enable pipeline parallel inference based on pipeline topology.
+
+    Args:
+        rank: The stage index of the current process
+        world_size: The total number of stages
+        llm: The LLM model shard instance
+        node: The node class instances for communication
+    """
+    assert world_size > 1, "Pipeline parallel inference requires at least 2 stages"
+
+    # Model runner owns sampler, model owns layers
+    model_runner: nn.Module = llm.llm_engine.model_executor.driver_worker.model_runner
+    model: nn.Module = model_runner.model
+
+    # Extract first and last layers (pre/post-hook to recv/send intermediate states)
+    first_layer: nn.Module = model.model.layers[0]
+    last_layer: nn.Module = model.model.layers[-1]
+
+    # Extract sampler (post-hook to recv/send outputs)
+    sampler: nn.Module = model_runner.sampler
+
+    # Don't relay outputs from stage with index -2->-1
+    relay = rank != world_size - 2
+
+    if rank == 0:  # First stage
+        # Send intermediate states to next stage (post-hook)
+        last_layer.register_forward_hook(partial(send_intermediate_states, node=node))
+        logger.info("Registered post-hook send_intermediate_states on last layer")
+
```

Comment:
```suggestion
        logger.debug("Registered post-hook send_intermediate_states on last layer")

```

I think that the granularity here could be up to debug and not info. But not strong opinion, whatever your prefer